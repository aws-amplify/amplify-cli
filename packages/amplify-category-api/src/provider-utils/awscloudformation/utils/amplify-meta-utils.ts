import { $TSAny, $TSMeta, $TSObject, AmplifyCategories, AmplifySupportedService, stateManager } from 'amplify-cli-core';
import _ from 'lodash';

export const authConfigHasApiKey = (authConfig?: $TSAny) => {
  if (!authConfig) {
    return false;
  }
  return (
    Array.of(authConfig.defaultAuthentication)
      .concat(authConfig.additionalAuthenticationProviders)
      .filter(auth => !!auth) // filter out undefined elements which can happen if there are no addtl auth providers
      .map(auth => auth.authenticationType)
      .findIndex(authType => authType === 'API_KEY') > -1
  );
};

export const checkIfAuthExists = () => {
  const amplifyMeta = stateManager.getMeta();
  let authResourceName;
  const authServiceName = AmplifySupportedService.COGNITO;
  const authCategoryName = AmplifyCategories.AUTH;

  if (amplifyMeta[authCategoryName] && Object.keys(amplifyMeta[authCategoryName]).length > 0) {
    const categoryResources = amplifyMeta[authCategoryName];
    Object.keys(categoryResources).forEach(resource => {
      if (categoryResources[resource].service === authServiceName) {
        authResourceName = resource;
      }
    });
  }

  return authResourceName;
};

// some utility functions to extract the AppSync API name and config from amplify-meta

export const getAppSyncAuthConfig = (projectMeta: $TSMeta) => {
  const entry = getAppSyncAmplifyMetaEntry(projectMeta);
  if (entry) {
    const value = entry[1] as $TSAny;
    return value && value.output ? value.output.authConfig : {};
  }
};

export const getAppSyncResourceName = (projectMeta: $TSMeta): string | undefined => {
  const entry = getAppSyncAmplifyMetaEntry(projectMeta);
  if (entry) {
    return entry[0];
  }
};

// project meta is the contents of amplify-meta.json
// typically retreived using context.amplify.getProjectMeta()
const getAppSyncAmplifyMetaEntry = (projectMeta: $TSMeta) => {
  return Object.entries(projectMeta[AmplifyCategories.API] || {}).find(
    ([, value]) => (value as $TSObject).service === AmplifySupportedService.APPSYNC,
  );
};
