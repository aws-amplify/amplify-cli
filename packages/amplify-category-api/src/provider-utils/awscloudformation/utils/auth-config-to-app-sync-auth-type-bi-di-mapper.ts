import {
  AppSyncAuthType,
  AppSyncAPIKeyAuthType,
  AppSyncCognitoUserPoolsAuthType,
  AppSyncOpenIDConnectAuthType,
  AppSyncLambdaAuthType,
} from 'amplify-headless-interface';
import _ from 'lodash';

/**
 * Converts the authConfig object that is returned by the AppSync walkthrough into the AppSyncAuthType defined by the AddApiRequest
 *
 * Used when transforming the walkthrough result into an AddApiRequest
 */
export const authConfigToAppSyncAuthType = (authConfig: any = {}): AppSyncAuthType => {
  return _.get(authConfigToAppSyncAuthTypeMap, authConfig.authenticationType, () => undefined)(authConfig);
};

/**
 * Converts an AppSyncAuthType object into the authConfig object that gets written to amplify-meta
 *
 * This conversion is necessary to ensure we are storing the authConfig in the same way as older projects
 * @param authType
 */
export const appSyncAuthTypeToAuthConfig = (authType?: AppSyncAuthType) => {
  if (!authType) return undefined;
  return _.get(appSyncAuthTypeToAuthConfigMap, authType.mode, () => undefined)(authType);
};

const authConfigToAppSyncAuthTypeMap: Record<string, (authConfig: any) => AppSyncAuthType> = {
  API_KEY: authConfig => ({
    mode: 'API_KEY',
    expirationTime: authConfig.apiKeyConfig.apiKeyExpirationDays,
    apiKeyExpirationDate: authConfig.apiKeyConfig?.apiKeyExpirationDate,
    keyDescription: authConfig.apiKeyConfig.description,
  }),
  AWS_IAM: () => ({
    mode: 'AWS_IAM',
  }),
  AMAZON_COGNITO_USER_POOLS: authConfig => ({
    mode: 'AMAZON_COGNITO_USER_POOLS',
    cognitoUserPoolId: authConfig.userPoolConfig.userPoolId,
  }),
  OPENID_CONNECT: authConfig => ({
    mode: 'OPENID_CONNECT',
    openIDProviderName: authConfig.openIDConnectConfig.name,
    openIDIssuerURL: authConfig.openIDConnectConfig.issuerUrl,
    openIDClientID: authConfig.openIDConnectConfig.clientId,
    openIDAuthTTL: authConfig.openIDConnectConfig.authTTL,
    openIDIatTTL: authConfig.openIDConnectConfig.iatTTL,
  }),
  AWS_LAMBDA: authConfig => ({
    mode: 'AWS_LAMBDA',
    lambdaFunction: authConfig.lambdaAuthorizerConfig.lambdaFunction,
    ttlSeconds: authConfig.lambdaAuthorizerConfig.ttlSeconds,
  }),
};

const appSyncAuthTypeToAuthConfigMap: Record<string, (authType: AppSyncAuthType) => any> = {
  API_KEY: (authType: AppSyncAPIKeyAuthType) => ({
    authenticationType: 'API_KEY',
    apiKeyConfig: {
      apiKeyExpirationDays: authType.expirationTime,
      apiKeyExpirationDate: authType?.apiKeyExpirationDate,
      description: authType.keyDescription,
    },
  }),
  AWS_IAM: () => ({
    authenticationType: 'AWS_IAM',
  }),
  AMAZON_COGNITO_USER_POOLS: (authType: AppSyncCognitoUserPoolsAuthType) => ({
    authenticationType: 'AMAZON_COGNITO_USER_POOLS',
    userPoolConfig: {
      userPoolId: authType.cognitoUserPoolId,
    },
  }),
  OPENID_CONNECT: (authType: AppSyncOpenIDConnectAuthType) => ({
    authenticationType: 'OPENID_CONNECT',
    openIDConnectConfig: {
      name: authType.openIDProviderName,
      issuerUrl: authType.openIDIssuerURL,
      clientId: authType.openIDClientID,
      authTTL: authType.openIDAuthTTL,
      iatTTL: authType.openIDIatTTL,
    },
  }),
  AWS_LAMBDA: (authType: AppSyncLambdaAuthType) => ({
    authenticationType: 'AWS_LAMBDA',
    lambdaAuthorizerConfig: {
      lambdaFunction: authType.lambdaFunction,
      ttlSeconds: authType.ttlSeconds,
    },
  }),
};
