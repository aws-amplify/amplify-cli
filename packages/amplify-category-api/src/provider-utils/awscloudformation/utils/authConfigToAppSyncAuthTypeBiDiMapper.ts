import {
  AppSyncAuthType,
  AppSyncAPIKeyAuthType,
  AppSyncCognitoUserPoolsAuthType,
  AppSyncOpenIDConnectAuthType,
} from 'amplify-headless-interface';
import _ from 'lodash';

export const authConfigToAppSyncAuthType = (authConfig: any = {}): AppSyncAuthType => {
  return _.get(authConfigToAppSyncAuthTypeMap, authConfig.authenticationType, () => undefined)(authConfig);
};

export const appSyncAuthTypeToAuthConfig = (authType?: AppSyncAuthType) => {
  if (!authType) return undefined;
  return _.get(appSyncAuthTypeToAuthConfigMap, authType.mode, () => undefined)(authType);
};

const authConfigToAppSyncAuthTypeMap: Record<string, (oldAuthObj: any) => AppSyncAuthType> = {
  API_KEY: oldAuthObj => ({
    mode: 'API_KEY',
    expirationTime: oldAuthObj.apiKeyConfig.apiKeyExpirationDays,
    keyDescription: oldAuthObj.apiKeyConfig.description,
  }),
  AWS_IAM: () => ({
    mode: 'AWS_IAM',
  }),
  AMAZON_COGNITO_USER_POOLS: oldAuthObj => ({
    mode: 'AMAZON_COGNITO_USER_POOLS',
    cognitoUserPoolId: oldAuthObj.userPoolConfig.userPoolId,
  }),
  OPENID_CONNECT: oldAuthObj => ({
    mode: 'OPENID_CONNECT',
    openIDProviderName: oldAuthObj.openIDConnectConfig.name,
    openIDIssuerURL: oldAuthObj.openIDConnectConfig.issuerUrl,
    openIDClientID: oldAuthObj.openIDConnectConfig.clientId,
    openIDAuthTTL: oldAuthObj.openIDConnectConfig.authTTL,
    openIDIatTTL: oldAuthObj.openIDConnectConfig.iatTTL,
  }),
};

const appSyncAuthTypeToAuthConfigMap: Record<string, (authType: AppSyncAuthType) => any> = {
  API_KEY: (authType: AppSyncAPIKeyAuthType) => ({
    authenticationType: 'API_KEY',
    apiKeyConfig: {
      apiKeyExpirationDays: authType.expirationTime,
      description: authType.keyDescription,
    },
  }),
  AWS_IAM: () => ({
    authenticationType: 'AWS_IAM',
  }),
  AMAZON_COGNITO_USER_POOLS: (authType: AppSyncCognitoUserPoolsAuthType) => ({
    authenticationType: 'AMAZON_COGNITO_USER_POOLS',
    userPoolConfig: {
      userPoolId: authType.cognitoUserPoolId,
    },
  }),
  OPENID_CONNECT: (authType: AppSyncOpenIDConnectAuthType) => ({
    authenticationType: 'OPENID_CONNECT',
    openIDConnectConfig: {
      name: authType.openIDProviderName,
      issuerUrl: authType.openIDIssuerURL,
      clientId: authType.openIDClientID,
      authTTL: authType.openIDAuthTTL,
      iatTTL: authType.openIDIatTTL,
    },
  }),
};
