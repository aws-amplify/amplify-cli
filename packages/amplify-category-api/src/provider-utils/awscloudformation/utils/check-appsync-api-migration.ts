import { $TSContext, AmplifyCategories, getMigrateResourceMessageForOverride } from 'amplify-cli-core';
import { printer, prompter } from 'amplify-prompts';
import { AppsyncApiInputState } from '../api-input-manager/appsync-api-input-state';
import { migrateResourceToSupportOverride } from './migrate-api-override-resource';

export const checkAppsyncApiResourceMigration = async (context: $TSContext, apiName: string, isUpdate: boolean): Promise<boolean> => {
  const cliState = new AppsyncApiInputState(apiName);
  if (!cliState.cliInputFileExists()) {
    printer.debug('cli-inputs.json doesnt exist');
    const headlessMigrate = context.input.options?.yes || context.input.options?.forcePush || context.input.options?.headless;
    if (headlessMigrate || (await prompter.yesOrNo(getMigrateResourceMessageForOverride(AmplifyCategories.API, apiName, isUpdate), true))) {
      // generate cli-inputs for migration from parameters.json
      await migrateResourceToSupportOverride(apiName);
      return true;
    }
    return false;
  }
  return true;
};
