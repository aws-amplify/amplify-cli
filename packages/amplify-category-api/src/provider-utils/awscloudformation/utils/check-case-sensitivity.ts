import fs from 'fs-extra';
import path from 'path';

export const checkCaseSensitivityIssue = async (context: any, category: string, resourceName: string) => {
  const projectBackendDirPath = context.amplify.pathManager.getBackendDirPath();

  const caseSensitivityConflict = async () => {
    const basePath = path.join(projectBackendDirPath, category);
    const filenames = await fs.readdir(basePath);
    const lowerCaseMatch = filenames.find(name => name.toLowerCase() === resourceName.toLowerCase());
    return lowerCaseMatch === resourceName ? false : lowerCaseMatch;
  };

  const conflict = await caseSensitivityConflict();
  if (conflict) {
    context.print.error(
      `Unable to create resource with name ${resourceName} since a resource named ${conflict} already exists. Amplify resource names are case-insensitive.`,
    );
    process.exit(1);
  }
};
