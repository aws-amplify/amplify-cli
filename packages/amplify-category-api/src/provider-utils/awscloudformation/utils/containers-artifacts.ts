import { Octokit } from '@octokit/rest';
import * as fs from 'fs-extra';
import inquirer from 'inquirer';
import * as path from 'path';
import { v4 as uuid } from 'uuid';
import { provider as cloudformationProviderName } from '../../../provider-utils/awscloudformation/aws-constants';
import { getContainers } from '../../../provider-utils/awscloudformation/docker-compose';
import Container from '../docker-compose/ecs-objects/container';
import { EcsStack } from '../ecs-apigw-stack';
import { API_TYPE, ResourceDependency } from '../../../provider-utils/awscloudformation/service-walkthroughs/containers-walkthrough';
import { getGitHubOwnerRepoFromPath } from '../../../provider-utils/awscloudformation/utils/github';
import { $TSAny, $TSContext, JSONUtilities, pathManager, readCFNTemplate } from 'amplify-cli-core';
import { DEPLOYMENT_MECHANISM } from '../base-api-stack';
import { setExistingSecretArns } from './containers/set-existing-secret-arns';
import { category } from '../../../category-constants';
import * as cdk from '@aws-cdk/core';

export const cfnFileName = (resourceName: string) => `${resourceName}-cloudformation-template.json`;

export type ApiResource = {
  category: string;
  resourceName: string;
  gitHubInfo?: {
    path: string;
    tokenSecretArn: string;
  };
  deploymentMechanism: DEPLOYMENT_MECHANISM;
  authName: string;
  restrictAccess: boolean;
  dependsOn: ResourceDependency[];
  environmentMap: Record<string, string>;
  categoryPolicies: $TSAny[];
  mutableParametersState: $TSAny;
  output?: Record<string, $TSAny>;
  apiType?: API_TYPE;
  exposedContainer?: { name: string; port: number };
};

type ExposedContainer = {
  name: string;
  port: number;
};

type ContainerArtifactsMetadata = {
  exposedContainer: ExposedContainer;
  pipelineInfo: { consoleUrl: string };
};

export async function generateContainersArtifacts(
  context: $TSContext,
  resource: ApiResource,
  askForExposedContainer: boolean = false,
): Promise<ContainerArtifactsMetadata> {
  const {
    providers: { [cloudformationProviderName]: provider },
  } = context.amplify.getProjectMeta();

  const { StackName: envName } = provider;

  const {
    category: categoryName,
    resourceName,
    gitHubInfo,
    deploymentMechanism,
    categoryPolicies = [],
    dependsOn,
    environmentMap,
    restrictAccess,
    apiType,
  } = resource;

  const backendDir = context.amplify.pathManager.getBackendDirPath();
  const resourceDir = path.normalize(path.join(backendDir, categoryName, resourceName));
  const srcPath = path.join(resourceDir, 'src');

  const { containersPorts, containers, isInitialDeploy, desiredCount, exposedContainer, secretsArns } = await processDockerConfig(
    context,
    resource,
    srcPath,
    askForExposedContainer,
  );
  const repositories = await context.amplify.executeProviderUtils(context, 'awscloudformation', 'describeEcrRepositories');

  const existingEcrRepositories: Set<string> = new Set(
    repositories
      .map(({ repositoryName }) => repositoryName)
      .filter(repositoryName => repositoryName.startsWith(`${envName}-${categoryName}-${resourceName}-`)),
  );

  const stack = new EcsStack(undefined, 'ContainersStack', {
    categoryName,
    apiName: resourceName,
    taskPorts: containersPorts,
    dependsOn,
    policies: categoryPolicies,
    taskEnvironmentVariables: environmentMap,
    gitHubSourceActionInfo: gitHubInfo,
    deploymentMechanism,
    containers,
    isInitialDeploy,
    desiredCount,
    restrictAccess,
    currentStackName: envName,
    apiType,
    exposedContainer,
    secretsArns,
    existingEcrRepositories,
  });

  const cfn = stack.toCloudFormation();

  JSONUtilities.writeJson(path.normalize(path.join(resourceDir, cfnFileName(resourceName))), cfn);

  return {
    exposedContainer,
    pipelineInfo: { consoleUrl: stack.getPipelineConsoleUrl(provider.Region) },
  };
}

export async function processDockerConfig(
  context: $TSContext,
  resource: ApiResource,
  srcPath: string,
  askForExposedContainer: boolean = false,
) {
  const {
    providers: { [cloudformationProviderName]: provider },
  } = context.amplify.getProjectMeta();

  const { StackName: envName } = provider;

  const { resourceName, gitHubInfo, deploymentMechanism, output, exposedContainer: exposedContainerFromMeta } = resource;

  const dockerComposeFileNameYaml = 'docker-compose.yaml';
  const dockerComposeFileNameYml = 'docker-compose.yml';
  const dockerfileFileName = 'Dockerfile';

  const containerDefinitionFileNames = [dockerComposeFileNameYaml, dockerComposeFileNameYml, dockerfileFileName];

  const containerDefinitionFiles: Record<string, string> = {};

  for await (const fileName of containerDefinitionFileNames) {
    switch (deploymentMechanism) {
      case DEPLOYMENT_MECHANISM.FULLY_MANAGED:
      case DEPLOYMENT_MECHANISM.SELF_MANAGED:
        const filePath = path.normalize(path.join(srcPath, fileName));

        if (fs.existsSync(filePath)) {
          containerDefinitionFiles[fileName] = fs.readFileSync(filePath).toString();
        }
        break;
      case DEPLOYMENT_MECHANISM.INDENPENDENTLY_MANAGED:
        const { path: repoUri, tokenSecretArn } = gitHubInfo;

        const { SecretString: gitHubToken } = await context.amplify.executeProviderUtils(context, 'awscloudformation', 'retrieveSecret', {
          secretArn: tokenSecretArn,
        });

        const octokit = new Octokit({ auth: gitHubToken });

        const { owner, repo, branch, path: pathInRepo } = getGitHubOwnerRepoFromPath(repoUri);

        try {
          const {
            data: { content, encoding },
          } = (await octokit.repos.getContent({
            owner,
            repo,
            ...(branch ? { ref: branch } : undefined), // only include branch if not undefined
            path: path.join(pathInRepo, fileName),
          })) as { data: { content?: string; encoding?: string } };

          containerDefinitionFiles[fileName] = Buffer.from(content, <BufferEncoding>encoding).toString('utf8');
        } catch (error) {
          const { status } = error;

          // It is ok if the file doesn't exist, we skip it
          if (status !== 404) {
            throw error;
          }
        }
        break;
      default: {
        const exhaustiveCheck: never = deploymentMechanism;
        throw new Error(`Unhandled type [${exhaustiveCheck}]`);
      }
    }
  }

  if (Object.keys(containerDefinitionFiles).length === 0) {
    throw new Error('No definition available (docker-compose.yaml / docker-compose.yml / Dockerfile)');
  }

  if (containerDefinitionFiles[dockerComposeFileNameYaml] && containerDefinitionFiles[dockerComposeFileNameYml]) {
    throw new Error('There should be only one docker-compose.yaml / docker-compose.yml)');
  }

  const composeContents = containerDefinitionFiles[dockerComposeFileNameYaml] || containerDefinitionFiles[dockerComposeFileNameYml];
  const { [dockerfileFileName]: dockerfileContents } = containerDefinitionFiles;

  const { buildspec, containers, service, secrets } = getContainers(composeContents, dockerfileContents);

  const containersPorts = containers.reduce(
    (acc, container) => acc.concat(container.portMappings.map(({ containerPort }) => containerPort)),
    <number[]>[],
  );

  const newContainersName = Array.from(new Set(containers.map(({ name }) => name)));

  let isInitialDeploy = Object.keys(output ?? {}).length === 0;
  const currentContainersSet = new Set(output?.ContainerNames?.split(','));
  // Service require all containers to exists
  isInitialDeploy = isInitialDeploy || newContainersName.some(newContainer => !currentContainersSet.has(newContainer));

  let exposedContainer: { name: string; port: number };

  const containersExposed = containers.filter(container => container.portMappings.length > 0);

  if (containersPorts.length === 0) {
    throw new Error('Service requires at least one exposed port');
  } else if (containersPorts.length > 1) {
    exposedContainer = await checkContainerExposed(containersExposed, exposedContainerFromMeta, askForExposedContainer);
  } else {
    exposedContainer = {
      name: containersExposed[0].name,
      port: containersExposed[0].portMappings[0].containerPort,
    };
  }

  fs.ensureDirSync(srcPath);
  fs.writeFileSync(path.join(srcPath, 'buildspec.yml'), buildspec);

  const secretsArns: Map<string, string> = new Map<string, string>();

  if ((await shouldUpdateSecrets(context, secrets)) || isInitialDeploy) {
    // Normalizes paths
    // Validate secrets file paths, existence and prefixes
    const errors = Object.entries(secrets).reduce((acc, [secretName, secretFilePath]) => {
      const baseDir = path.isAbsolute(secretFilePath) ? '' : srcPath;
      const normalizedFilePath = path.normalize(path.join(baseDir, secretFilePath));

      secrets[secretName] = normalizedFilePath;

      let canRead = true;

      try {
        const fd = fs.openSync(normalizedFilePath, 'r');

        fs.closeSync(fd);
      } catch (err) {
        canRead = false;
      }

      if (!canRead) {
        acc.push(`Secret file "${secretFilePath}" can't be read.`);
        return acc;
      }

      const basename = path.basename(normalizedFilePath);
      const hasCorrectPrefix = basename.startsWith('.secret-');

      if (!hasCorrectPrefix) {
        acc.push(`Secret file "${secretFilePath}" doesn't start with the ".secret-" prefix.`);
        return acc;
      }

      const isInsideSrc = normalizedFilePath.startsWith(path.join(srcPath, path.sep));
      if (isInsideSrc) {
        acc.push(`Secret file "${secretFilePath}" should not be inside the "src" folder. The "src" folder will be uploaded to S3.`);
        return acc;
      }

      return acc;
    }, <string[]>[]);

    if (errors.length > 0) {
      throw new Error(['Error(s) in secret file(s):'].concat(errors).join('\n'));
    }

    for await (const entries of Object.entries(secrets)) {
      const [secretName, secretFilePath] = entries;

      const contents = fs.readFileSync(secretFilePath).toString();

      const ssmSecretName = `${envName}-${resourceName}-${secretName}`;

      const { ARN: secretArn } = await context.amplify.executeProviderUtils(context, 'awscloudformation', 'upsertSecretValue', {
        secret: contents,
        description: `Secret for ${resourceName}`,
        name: ssmSecretName,
        version: uuid(),
      });

      const [prefix,] = secretArn.toString().split(ssmSecretName);
      const secretArnRef = cdk.Fn.join('', [
        prefix,
        cdk.Fn.ref('rootStackName'),
        '-',
        resourceName,
        '-',
        secretName,
      ]);

      secretsArns.set(secretName, secretArnRef);
    }
  } else {
    const { cfnTemplate } = readCFNTemplate(path.join(pathManager.getBackendDirPath(), category, resourceName, cfnFileName(resourceName)));
    setExistingSecretArns(secretsArns, cfnTemplate);
  }

  const desiredCount = service?.replicas ?? 1; // TODO: 1 should be from meta (HA setting)

  return {
    containersPorts,
    containers,
    isInitialDeploy,
    desiredCount,
    exposedContainer,
    secretsArns,
  };
}

async function shouldUpdateSecrets(context: $TSContext, secrets: Record<string, string>): Promise<boolean> {
  const hasSecrets = Object.keys(secrets).length > 0;

  if (!hasSecrets || context.exeInfo.inputParams.yes) {
    return false;
  }

  const { update_secrets } = await inquirer.prompt({
    name: 'update_secrets',
    type: 'confirm',
    message: 'Secret configuration detected. Do you wish to store new values in the cloud?',
    default: false,
  });

  return update_secrets;
}

async function checkContainerExposed(
  containersExposed: Container[],
  exposedContainerFromMeta: { name: string; port: number } = { name: '', port: 0 },
  askForExposedContainer: boolean = false,
): Promise<{ name: string; port: number }> {
  const containerExposed = containersExposed.find(container => container.name === exposedContainerFromMeta.name);

  if (!askForExposedContainer && containerExposed?.portMappings.find(port => port.containerPort === exposedContainerFromMeta.port)) {
    return { ...exposedContainerFromMeta };
  } else {
    const choices: { name: string; value: Container }[] = containersExposed.map(container => ({ name: container.name, value: container }));

    const { containerToExpose } = await inquirer.prompt({
      message: 'Select which container is the entrypoint',
      name: 'containerToExpose',
      type: 'list',
      choices,
    });

    return {
      name: containerToExpose.name,
      port: containerToExpose.portMappings[0].containerPort,
    };
  }
}
