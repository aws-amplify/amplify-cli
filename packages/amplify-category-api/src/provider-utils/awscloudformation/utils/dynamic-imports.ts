import * as path from 'path';

export const serviceMetadataFor = async (service: string) =>
  (await import(path.join('..', '..', 'supported-services'))).supportedServices[service];

export const datasourceMetadataFor = async (datasource: string) =>
  (await import(path.join('..', '..', 'supported-datasources'))).supportedDatasources[datasource];

export const getServiceWalkthrough = async (walkthroughFilename: string) =>
  (await import(path.join('..', 'service-walkthroughs', walkthroughFilename))).serviceWalkthrough;
