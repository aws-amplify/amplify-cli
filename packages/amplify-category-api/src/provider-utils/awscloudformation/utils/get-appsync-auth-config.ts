import { AppsyncApiInputState } from '../api-input-manager/appsync-api-input-state';
import { appSyncAuthTypeToAuthConfig } from './auth-config-to-app-sync-auth-type-bi-di-mapper';

/**
 *
 * @param resourceName
 * @returns authConfig
 */
export const getAuthConfig = async (resourceName: string) => {
  const cliState = new AppsyncApiInputState(resourceName);
  if (cliState.cliInputFileExists()) {
    const appsyncInputs = cliState.getCLIInputPayload().serviceConfiguration;
    return {
      defaultAuthentication: appSyncAuthTypeToAuthConfig(appsyncInputs.defaultAuthType),
      additionalAuthenticationProviders: (appsyncInputs.additionalAuthTypes || []).map(appSyncAuthTypeToAuthConfig),
    };
  }
};
