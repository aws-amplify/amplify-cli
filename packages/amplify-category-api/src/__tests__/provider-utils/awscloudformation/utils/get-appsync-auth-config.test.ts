import { getAuthConfig } from '../../../../provider-utils/awscloudformation/utils/get-appsync-auth-config';

const getCLIInputPayload_mock = jest
  .fn()
  .mockReturnValueOnce({
    version: 1,
    serviceConfiguration: {
      apiName: 'authv2migration1',
      serviceName: 'AppSync',
      gqlSchemaPath: 'mock/schema.graphql',
      defaultAuthType: {
        mode: 'AWS_IAM',
      },
      conflictResolution: {},
      additionalAuthTypes: [],
    },
  })
  .mockReturnValueOnce({
    version: 1,
    serviceConfiguration: {
      apiName: 'authv2migration1',
      serviceName: 'AppSync',
      gqlSchemaPath: 'mock/schema.graphql',
      defaultAuthType: {
        mode: 'AWS_IAM',
      },
      conflictResolution: {},
      additionalAuthTypes: [
        {
          mode: 'API_KEY',
          expirationTime: 7,
          keyDescription: '',
        },
      ],
    },
  });

jest.mock('../../../../provider-utils/awscloudformation/api-input-manager/appsync-api-input-state.ts', () => {
  return {
    AppsyncApiInputState: jest.fn().mockImplementation(() => {
      return {
        getCLIInputPayload: getCLIInputPayload_mock,
        cliInputFileExists: jest.fn().mockReturnValue(true),
      };
    }),
  };
});

test('test function with default auth config', async () => {
  expect(await getAuthConfig('mockapiResource')).toMatchSnapshot();
});

test('test function with default and additional auth config', async () => {
  expect(await getAuthConfig('mockapiResource')).toMatchSnapshot();
});
