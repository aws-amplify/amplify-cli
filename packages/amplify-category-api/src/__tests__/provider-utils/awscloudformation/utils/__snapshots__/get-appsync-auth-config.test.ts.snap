// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test function with default and additional auth config 1`] = `
Object {
  "additionalAuthenticationProviders": Array [
    Object {
      "apiKeyConfig": Object {
        "apiKeyExpirationDate": undefined,
        "apiKeyExpirationDays": 7,
        "description": "",
      },
      "authenticationType": "API_KEY",
    },
  ],
  "defaultAuthentication": Object {
    "authenticationType": "AWS_IAM",
  },
}
`;

exports[`test function with default auth config 1`] = `
Object {
  "additionalAuthenticationProviders": Array [],
  "defaultAuthentication": Object {
    "authenticationType": "AWS_IAM",
  },
}
`;
